File: http patch/Scala/fundservice/fund/src/main/scala/fund/DomainModel.scala
package fund

import fund.PatchInstructionService._
import fund.common.Pipeline._
import fund.common.StringExtensions._
import fund.common.SupportException

import scala.collection.mutable.ListBuffer

object DomainModel {
  val INVALID_FUND_NAME_MESSAGE = "Invalid fund name"
  val INVALID_FUND_TYPE_MESSAGE = "Invalid fund type"

  class PatchFundDto(val operation: String, val path: String, val value: String) { }

  case class Fund(fundId: Int, fundName: String, fundType: String)

  def validateFundName(fund: Fund): Result[Fund] = {
    fund.fundName match {
      case fundName if fundName.isNullOrEmpty => Result(INVALID_FUND_NAME_MESSAGE, fund)
      case _ => Result(data = fund)
    }
  }

  def validateFundType(fund: Fund): Result[Fund] = {
    fund.fundType match {
      case fundType if fundType.isNullOrEmpty => Result(INVALID_FUND_TYPE_MESSAGE, fund)
      case _ => Result(data = fund)
    }
  }

  def updateFundName(fundName: String) = {
    val apply = (data: Fund) => {
      Result(data = Fund(data.fundId, fundName, data.fundType))
    }: Result[Fund]
    apply
  }

  def updateFundType(fundType: String) = {
    val apply = (data: Fund) => {
      Result(data = Fund(data.fundId, data.fundName, fundType))
    }: Result[Fund]
    apply
  }

  def patchFund(id: Int, instructions: List[PatchFundDto], get: ((Int) => Fund), update: ((Fund) => Result[Fund])): Unit = {
    val fund = get(id)
    val funcs = ListBuffer[((Fund) => Result[Fund])](validateFundName, validateFundType)

    instructions.foreach(instruction => {
      instruction match {
        case instruction if instruction.path == "fundType" => funcs += patchFundType(fund, instruction)
        case instruction if instruction.path == "fundName" => funcs += patchFundName(fund, instruction)
      }
    })

    runUntilFirstFault(fund, funcs.toList) match {
      case result if result.success() => update(result.data)
      case result => throw new SupportException(result.error)
    }
  }
}

File: http patch/C#/FundService/FundService/DomainModel.cs
using System;
using System.Collections.Generic;
using System.Linq;
using FundService.Common;
using static FundService.Common.Pipeline;

namespace FundService
{
	public static class DomainModel
	{
		public const string INVALID_FUND_NAME_MESSAGE = "Invalid fund name";
		public const string INVALID_FUND_TYPE_MESSAGE = "Invalid fund type";
        public const string FUND_TYPE = "fundType";
        public const string FUND_NAME = "fundName";


        public static Result<Fund> ValidateFundName(Fund fund) {
			if (string.IsNullOrWhiteSpace(fund.FundName)) return new Result<Fund>(INVALID_FUND_NAME_MESSAGE, fund);
			return new Result<Fund>(data: fund);
		}

		public static Result<Fund> ValidateFundType(Fund fund) {
			if (string.IsNullOrWhiteSpace(fund.FundType)) return new Result<Fund>(INVALID_FUND_TYPE_MESSAGE, fund);
			return new Result<Fund>(data: fund);
		}

		public static Func<Fund, Result<Fund>> UpdateFundName(string fundName) {
			return (fund) => {
				return new Result<Fund>(data: new Fund(fund.Id, fundName, fund.FundType));
			};
		}

		public static Func<Fund, Result<Fund>> UpdateFundType(string fundType) {
			return (fund) => {
				return new Result<Fund>(data: new Fund(fund.Id, fund.FundName, fundType));
			};
		}

		public static void PatchFund(int id, IEnumerable<PatchFundDto> instructions, Func<int, Fund> getFund, Func<Fund, Result<Fund>> updateFund) {
			var fund = getFund(id);
			var funcs = new List<Func<Fund, Result<Fund>>>();

			funcs.Add(ValidateFundName);
			funcs.Add(ValidateFundType);

			instructions.ToList().ForEach(instruction => {
				switch (instruction.Path) {
					case "fundName":
						funcs.Add(UpdateFundName(fund.FundName));
						break;
					case "fundType":
						funcs.Add(UpdateFundType(fund.FundType));
					break;
				}
			});

			var result = RunUntilFirstFault(fund, funcs);
			if (result.Success) updateFund(result.Data);
			else throw new SupportException(result.Error);
		}
	}
}
